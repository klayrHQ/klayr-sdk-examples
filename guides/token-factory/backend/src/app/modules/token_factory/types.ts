import { JSONObject } from 'klayr-sdk';

export type TokenID = Buffer;

export interface ModuleConfig {
	maxNameLength: number;
	maxSymbolLength: number;
	maxTotalSupply: bigint;
	minAmountToMint: bigint;
	maxAmountToMint: bigint;
	chainID: Buffer;
	createTokenFee: bigint;
}

export type ModuleConfigJSON = JSONObject<ModuleConfig>;
