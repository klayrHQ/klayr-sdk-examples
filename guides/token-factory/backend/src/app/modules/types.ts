import { JSONObject } from 'lisk-sdk';

export interface ModuleConfig {
	maxNameLength: number;
	maxSymbolLength: number;
	maxTotalSupply: number;
}

export type ModuleConfigJSON = JSONObject<ModuleConfig>;
