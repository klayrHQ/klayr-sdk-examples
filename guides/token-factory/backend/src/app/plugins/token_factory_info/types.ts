export interface TokenFactoryInfoPluginConfig {
	syncInterval: number;
	chainID: string;
}

export interface NewTokenEvent {
	owner: Buffer;
	tokenID: Buffer;
	name: string;
	symbol: string;
	totalSupply: bigint;
	height: number;
}

export interface TokenIDCounter {
	tokenID: number;
}

export interface Height {
	height: number;
}

export interface ApiEventResult {
	data: string;
	height: number;
	module: string;
	name: string;
}
