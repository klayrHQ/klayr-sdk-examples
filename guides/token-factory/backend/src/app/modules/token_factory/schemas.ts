export const mintSchema = {
	$id: 'token_factory/mintParams',
	title: 'Mints extra tokens to a recipient',
	type: 'object',
	properties: {
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		amount: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		recipient: {
			dataType: 'bytes',
			fieldNumber: 3,
			format: 'klayr32',
		},
	},
	required: ['amount', 'recipient'],
};

export const createTokenSchema = {
	$id: 'token_factory/createToken-params',
	title: 'CreateTokenCommand transaction parameter for the Token Factory module',
	type: 'object',
	required: ['name', 'symbol', 'totalSupply'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
			minLength: 1,
			maxLength: 256,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 2,
			minLength: 1,
			maxLength: 256,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
	},
};

export const configSchema = {
	$id: '/token_factory/config',
	type: 'object',
	properties: {
		maxNameLength: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		maxSymbolLength: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		maxTotalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		minAmountToMint: {
			dataType: 'uint64',
			fieldNumber: 4,
		},
		maxAmountToMint: {
			dataType: 'uint64',
			fieldNumber: 5,
		},
		createTokenFee: {
			dataType: 'uint64',
			fieldNumber: 6,
		},
	},
	required: [
		'maxNameLength',
		'maxSymbolLength',
		'maxTotalSupply',
		'minAmountToMint',
		'maxAmountToMint',
		'createTokenFee',
	],
};
