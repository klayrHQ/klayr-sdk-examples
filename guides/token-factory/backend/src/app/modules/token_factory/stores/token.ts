import { BaseStore } from 'klayr-sdk';

export interface TokenStoreData {
	name: string;
	symbol: string;
	totalSupply: bigint;
}

export const tokenStoreSchema = {
	$id: '/token_factory/token',
	type: 'object',
	required: ['name', 'symbol', 'totalSupply'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 2,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
	},
};

export class TokenStore extends BaseStore<TokenStoreData> {
	public schema = tokenStoreSchema;
}
