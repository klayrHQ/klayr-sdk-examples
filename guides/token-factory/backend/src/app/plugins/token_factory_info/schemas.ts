export const configSchema = {
	$id: '/plugins/token_factory/config',
	type: 'object',
	properties: {
		syncInterval: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		chainID: {
			dataType: 'string',
			fieldNumber: 2,
		},
	},
	required: ['syncInterval', 'chainID'],
	default: {
		syncInterval: 20000, // milliseconds
		chainID: '12345678',
	},
};

export const mintEventSchema = {
	$id: 'token_factory/mintEvent',
	type: 'object',
	properties: {
		address: {
			dataType: 'bytes',
			fieldNumber: 1,
			format: 'klayr32',
		},
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 2,
		},
		amount: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		result: {
			dataType: 'uint32',
			fieldNumber: 4,
		},
	},
	required: ['address', 'tokenID', 'amount', 'result'],
};

export const newTokenEventSchema = {
	$id: 'token_factory/new_token',
	title: 'New token created with the createToken command',
	type: 'object',
	required: ['owner', 'name', 'symbol', 'totalSupply', 'height'],
	properties: {
		owner: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 2,
		},
		name: {
			dataType: 'string',
			fieldNumber: 3,
			minLength: 1,
			maxLength: 256,
		},
		symbol: {
			dataType: 'string',
			fieldNumber: 4,
			minLength: 1,
			maxLength: 256,
		},
		totalSupply: {
			dataType: 'uint64',
			fieldNumber: 5,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 6,
		},
	},
};

export const heightSchema = {
	$id: '/token_factory/height',
	type: 'object',
	required: ['height'],
	properties: {
		height: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
	},
};
