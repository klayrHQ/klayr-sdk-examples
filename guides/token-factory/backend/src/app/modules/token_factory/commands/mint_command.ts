/* eslint-disable class-methods-use-this */

import {
	BaseCommand,
	CommandVerifyContext,
	CommandExecuteContext,
	VerificationResult,
	VerifyStatus,
	TokenMethod,
} from 'klayr-sdk';
import { mintSchema } from '../schemas';
import { TokenID } from '../types';
import { TokenStore } from '../stores/token';
import { failWithLog } from '../utils';
import { TokenFactoryMethod } from '../method';

export interface MintParams {
	tokenID: TokenID;
	amount: bigint;
	recipient: string;
}

export class MintCommand extends BaseCommand {
	private _minAmountToMint!: bigint;
	private _maxAmountToMint!: bigint;
	private _tokenFactoryMethod!: TokenFactoryMethod;
	private _tokenMethod!: TokenMethod;

	public schema = mintSchema;

	public addDependencies(args: {
		tokenFactoryMethod: TokenFactoryMethod;
		tokenMethod: TokenMethod;
	}) {
		this._tokenFactoryMethod = args.tokenFactoryMethod;
		this._tokenMethod = args.tokenMethod;
	}

	public async init(args: { minAmountToMint: bigint; maxAmountToMint: bigint }): Promise<void> {
		this._minAmountToMint = args.minAmountToMint;
		this._maxAmountToMint = args.maxAmountToMint;
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(context: CommandVerifyContext<MintParams>): Promise<VerificationResult> {
		const { amount, tokenID } = context.params;

		const owner = await this._tokenFactoryMethod.getOwnerOfToken<MintParams>(context, tokenID);
		if (!owner) {
			return failWithLog<MintParams>(context, `Invalid tokenID: ${tokenID}`);
		}

		if (!owner.address.equals(context.transaction.senderAddress)) {
			return failWithLog<MintParams>(context, `Sender is not the token creator`);
		}

		if (amount > this._maxAmountToMint || amount < this._minAmountToMint) {
			return failWithLog<MintParams>(
				context,
				`Amount can not be lower than ${this._minAmountToMint} or greater than ${this._maxAmountToMint}`,
			);
		}

		return { status: VerifyStatus.OK };
	}

	public async execute(context: CommandExecuteContext<MintParams>): Promise<void> {
		context.logger.info('EXECUTE Mint');
		const { recipient, tokenID, amount } = context.params;
		const tokenStore = this.stores.get(TokenStore);

		await this._tokenMethod.mint(
			context.getMethodContext(),
			Buffer.from(recipient),
			Buffer.from(tokenID),
			BigInt(amount),
		);

		const token = await tokenStore.get(context, tokenID);
		token.totalSupply += BigInt(amount);
		await tokenStore.set(context, tokenID, token);
	}
}
