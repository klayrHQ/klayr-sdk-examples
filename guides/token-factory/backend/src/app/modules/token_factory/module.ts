/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */

import { validator } from '@klayr/validator';
import {
	BaseModule,
	FeeMethod,
	ModuleInitArgs,
	ModuleMetadata,
	TokenMethod,
	utils,
} from 'klayr-sdk';
import { BatchTransferCommand } from './commands/batch_transfer_command';
import { ModuleConfig, ModuleConfigJSON } from './types';
import { BurnCommand } from './commands/burn_command';
import { CreateTokenCommand } from './commands/create_token_command';
import { MintCommand } from './commands/mint_command';
import { TokenFactoryEndpoint } from './endpoint';
import { NewTokenEvent } from './events/new_token';
import { InternalMethod } from './internal_methods';
import { TokenFactoryMethod } from './method';
import { configSchema } from './schemas';
import { CounterStore } from './stores/counter';
import { OwnerStore } from './stores/owner';
import { TokenStore } from './stores/token';
import { getModuleConfig } from './utils';

export const defaultConfig = {
	maxNameLength: 30,
	maxSymbolLength: 5,
	maxTotalSupply: BigInt(1e18), // max: 9223372036854775807 = 9e18
	minAmountToMint: BigInt(1000),
	maxAmountToMint: BigInt(1e6) * BigInt(1e8),
	createTokenFee: BigInt(100_000),
	minAmountToBurn: BigInt(1000),
};

export class TokenFactoryModule extends BaseModule {
	private readonly _internalMethod = new InternalMethod(this.stores, this.events);

	private _createTokenCommand = new CreateTokenCommand(this.stores, this.events);
	private _batchTransferCommand = new BatchTransferCommand(this.stores, this.events);
	private _mintCommand = new MintCommand(this.stores, this.events);
	private _burnCommand = new BurnCommand(this.stores, this.events);

	private _moduleConfig!: ModuleConfig;
	private _tokenMethod!: TokenMethod;
	private _feeMethod!: FeeMethod;

	public endpoint = new TokenFactoryEndpoint(this.stores, this.offchainStores);
	public tokenFactoryMethod = new TokenFactoryMethod(this.stores, this.events);
	public commands = [
		this._createTokenCommand,
		this._mintCommand,
		this._burnCommand,
		this._batchTransferCommand,
	];

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(TokenStore, new TokenStore(this.name, 0));
		this.stores.register(CounterStore, new CounterStore(this.name, 1));
		this.stores.register(OwnerStore, new OwnerStore(this.name, 2));

		this.events.register(NewTokenEvent, new NewTokenEvent(this.name));
	}

	public addDependencies(tokenMethod: TokenMethod, feeMethod: FeeMethod) {
		this._tokenMethod = tokenMethod;
		this._feeMethod = feeMethod;
		this._createTokenCommand.addDependencies({
			internalMethod: this._internalMethod,
			tokenMethod: this._tokenMethod,
			feeMethod: this._feeMethod,
		});
		this._mintCommand.addDependencies({
			tokenFactoryMethod: this.tokenFactoryMethod,
			tokenMethod: this._tokenMethod,
		});
		this._burnCommand.addDependencies({
			tokenFactoryMethod: this.tokenFactoryMethod,
			tokenMethod: this._tokenMethod,
		});
		this._batchTransferCommand.addDependencies({
			tokenMethod: this._tokenMethod,
		});
	}

	public metadata(): ModuleMetadata {
		return {
			...this.baseMetadata(),
			commands: this.commands.map(command => ({
				name: command.name,
				params: command.schema,
			})),
			events: this.events.values().map(v => ({
				name: v.name,
				data: v.schema,
			})),
			endpoints: [],
			assets: [],
		};
	}

	// Lifecycle hooks
	public async init(args: ModuleInitArgs): Promise<void> {
		// Get the module config defined in the config.json of the node
		const { moduleConfig, genesisConfig } = args;
		// Overwrite the default module config with values from config.json, if set
		const config = utils.objects.mergeDeep({}, defaultConfig, moduleConfig) as ModuleConfigJSON;
		// Validate the config with the config schema

		validator.validate<ModuleConfigJSON>(configSchema, config);

		this._moduleConfig = getModuleConfig(config, genesisConfig);
		this._internalMethod.init(this._moduleConfig.chainID);
		this._createTokenCommand.init(this._moduleConfig).catch(err => {
			console.log('Error: ', err);
		});
		this._mintCommand
			.init({
				minAmountToMint: this._moduleConfig.minAmountToMint,
				maxAmountToMint: this._moduleConfig.maxAmountToMint,
			})
			.catch(err => {
				console.log('Error: ', err);
			});
		this._burnCommand
			.init({
				minAmountToBurn: this._moduleConfig.minAmountToBurn,
			})
			.catch(err => {
				console.log('Error: ', err);
			});
	}

	// public async insertAssets(_context: InsertAssetContext) {
	// 	// initialize block generation, add asset
	// }

	// public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
	// 	// verify block
	// }

	// Lifecycle hooks
	// public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
	// verify transaction will be called multiple times in the transaction pool
	// return { status: VerifyStatus.OK };
	// }

	// public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
	// }

	// public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {

	// }
	// public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {

	// }

	// public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {

	// }
}
