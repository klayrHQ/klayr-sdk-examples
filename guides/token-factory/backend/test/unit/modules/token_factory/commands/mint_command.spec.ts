import { MintCommand, MintParams } from '@app/modules/token_factory/commands/mint_command';
import { TokenFactoryModule } from '@app/modules/token_factory/module';
import { createTokenSchema, mintSchema } from '@app/modules/token_factory/schemas';
import {
	TokenID,
	createCreateTokenCtx,
	createMintCtx,
	createSampleTransaction,
} from '@test/helpers';
import { CommandExecuteContext, Transaction, VerifyStatus, chain, codec, db } from 'klayr-sdk';
import { utils } from '@klayr/cryptography';
import {
	CreateTokenCommand,
	CreateTokenParams,
} from '@app/modules/token_factory/commands/create_token_command';
import { ModuleConfig } from '@app/modules/token_factory/types';
import { TokenStore } from '@app/modules/token_factory/stores/token';

describe('MintCommand', () => {
	const initConfig = {
		minAmountToMint: BigInt(1000),
		maxAmountToMint: BigInt(1e6) * BigInt(1e8),
		chainID: Buffer.from('12345678'),
		createTokenFee: BigInt(100_000),
	};
	const tokenID = new TokenID(1).toBuffer();
	const recipient = utils.getRandomBytes(20);

	const defaultValidParams = codec.encode(mintSchema, {
		tokenID,
		amount: BigInt(1000) * BigInt(1e8),
		recipient,
	});

	const mockMint = jest.fn();
	const mockInitialize = jest.fn();
	const mockPayFee = jest.fn();

	let mintCommand: MintCommand;
	let createCommand: CreateTokenCommand;
	let stateStore: any;
	let tokenStore: TokenStore;

	beforeEach(async () => {
		const { minAmountToMint, maxAmountToMint } = initConfig;
		const tokenFactory = new TokenFactoryModule();

		mintCommand = new MintCommand(tokenFactory.stores, tokenFactory.events);
		mintCommand.addDependencies({ tokenMethod: { mint: mockMint } as any });
		await mintCommand.init({ minAmountToMint, maxAmountToMint });

		createCommand = new CreateTokenCommand(tokenFactory.stores, tokenFactory.events);
		createCommand.addDependencies({
			tokenMethod: { mint: mockMint, initializeToken: mockInitialize },
			feeMethod: { payFee: mockPayFee },
		} as any);
		await createCommand.init(initConfig as ModuleConfig);

		stateStore = new chain.StateStore(new db.InMemoryDatabase());
		tokenStore = tokenFactory.stores.get(TokenStore);
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(mintCommand.name).toEqual('mint');
		});

		it('should have valid schema', () => {
			expect(mintCommand.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
			beforeEach(async () => {
				const defaultToken = {
					name: 'The real pepe coin',
					symbol: 'PEPE',
					totalSupply: BigInt(1e4),
				};
				const defaultValidParams = codec.encode(createTokenSchema, defaultToken);
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, CreateTokenCommand.name),
				);
				const context = createCreateTokenCtx(stateStore, transaction, 'execute');

				await expect(
					createCommand.execute(context as CommandExecuteContext<CreateTokenParams>),
				).resolves.toBeUndefined();
			});

			it('should throw when `amount` is too high', async () => {
				const recipient = utils.getRandomBytes(20);
				const paramWithInvalidamount = codec.encode(mintSchema, {
					tokenID,
					amount: BigInt(1e6) * BigInt(1e8) + BigInt(1),
					recipient,
				});
				const transaction = new Transaction(
					createSampleTransaction(paramWithInvalidamount, MintCommand.name),
				);
				const context = createMintCtx(stateStore, transaction, 'verify');

				const result = await mintCommand.verify(context);
				expect(result.status).toBe(VerifyStatus.FAIL);
			});

			it('should throw when `amount` is too low', async () => {
				const recipient = utils.getRandomBytes(20);
				const paramWithInvalidamount = codec.encode(mintSchema, {
					tokenID,
					amount: BigInt(1000) - BigInt(1),
					recipient,
				});
				const transaction = new Transaction(
					createSampleTransaction(paramWithInvalidamount, MintCommand.name),
				);
				const context = createMintCtx(stateStore, transaction, 'verify');

				const result = await mintCommand.verify(context);
				expect(result.status).toBe(VerifyStatus.FAIL);
			});

			it('should throw when sender is not the token owner / creator', async () => {
				const recipient = utils.getRandomBytes(21);
				const differentSender = 'ce2e1c458bb616e2717aa48f7392b39b1bc01af7297755db7b6d1db605fa0ef7';
				const validParams = codec.encode(mintSchema, {
					tokenID,
					amount: BigInt(1000) + BigInt(1),
					recipient,
				});
				const transaction = new Transaction(
					createSampleTransaction(validParams, MintCommand.name, differentSender),
				);
				const context = createMintCtx(stateStore, transaction, 'verify');

				const result = await mintCommand.verify(context);

				expect(result.status).toBe(VerifyStatus.FAIL);
				expect(result.error).toEqual(new Error('Sender is not the token creator'));
			});

			it('should be OK when params are correct', async () => {
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, MintCommand.name),
				);
				const context = createMintCtx(stateStore, transaction, 'verify');

				const result = await mintCommand.verify(context);
				expect(result.status).toBe(VerifyStatus.OK);
			});
		});
	});

	describe('execute', () => {
		describe('valid cases', () => {
			beforeEach(async () => {
				const defaultToken = {
					name: 'The real pepe coin',
					symbol: 'PEPE',
					totalSupply: BigInt(1e4),
				};
				const defaultValidParams = codec.encode(createTokenSchema, defaultToken);
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, CreateTokenCommand.name),
				);
				const context = createCreateTokenCtx(stateStore, transaction, 'execute');

				await expect(
					createCommand.execute(context as CommandExecuteContext<CreateTokenParams>),
				).resolves.toBeUndefined();
			});

			it('mint function should have been called and updated `totalSupply` ', async () => {
				const transaction = new Transaction(
					createSampleTransaction(defaultValidParams, MintCommand.name),
				);
				const context = createMintCtx(stateStore, transaction, 'execute');

				await expect(
					mintCommand.execute(context as CommandExecuteContext<MintParams>),
				).resolves.toBeUndefined();

				const token = await tokenStore.get(context, tokenID);

				// initial supply + minted amount
				expect(token.totalSupply).toBe(BigInt(1e4) + BigInt(1000) * BigInt(1e8));
				expect(mockMint).toHaveBeenCalledTimes(2);
			});
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
