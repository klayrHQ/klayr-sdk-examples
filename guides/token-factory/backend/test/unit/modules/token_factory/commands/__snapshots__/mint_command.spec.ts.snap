// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MintCommand constructor should have valid schema 1`] = `
{
  "$id": "token_factory/mintParams",
  "properties": {
    "amount": {
      "dataType": "uint64",
      "fieldNumber": 2,
    },
    "recipient": {
      "dataType": "bytes",
      "fieldNumber": 3,
      "format": "klayr32",
    },
    "tokenID": {
      "dataType": "bytes",
      "fieldNumber": 1,
    },
  },
  "required": [
    "amount",
    "recipient",
  ],
  "title": "Mints extra tokens to a recipient",
  "type": "object",
}
`;
